[1162] - Support database name and table name arguments to DeltaTable…

mmengarelli opened the pull request:
Added functionality in DeltaTable::isDeltaTable() to lookup by table name

Added unit test to DeltaTableSuite.scala for said functionality

This resolves #[1162](https://github.com/delta-io/delta/issues/1162)

zsxwing commented:
@mmengarelli Could you resolve the conflicts?

mmengarelli commented:
@zsxwing Yes I can resolve these

allisonport-db commented:
So overall, if we want to keep our API consistent we should update both `def isDeltaTable(sparkSession: SparkSession, identifier: String): Boolean` and `def isDeltaTable(identifier: String): Boolean` since we support these:
```
def forName(sparkSession: SparkSession, tableName: String): DeltaTable
def forName(tableOrViewName: String): DeltaTable
def forPath(path: String): DeltaTable
def forPath(sparkSession: SparkSession, path: String): DeltaTable
```

allisonport-db commented:
> @allisonport-db `isDeltaTable(identifier: String)` delegates to `isDeltaTable(sparkSession: SparkSession, identifier: String)` after attempting to find the SparkSession, so the functionality is covered in both. Can you confirm if this is what you are suggesting or is there something else?
> 
> I do realize that the javadoc for `isDeltaTable(identifier: String)` does need updating though.

Oops, yup my mistake :) Sorry bout that, I'll take a look at the code today. Just update the java doc then.

mmengarelli commented:
No worries @allisonport-db, just wanted to confirm. I will update the other javadoc then push

allisonport-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -17,21 +17,20 @@
 package io.delta.tables
 
 import java.sql.Timestamp
-
```
nit: add these new lines back (and 2 below)

mmengarelli replied:
Roger

allisonport-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -17,21 +17,20 @@
 import org.apache.spark.sql.catalyst.TableIdentifier
 import org.apache.spark.sql.types.StructType
 
+import scala.util.Try
```
nit: this goes with the other scala library imports

mmengarelli replied:
Roger

allisonport-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -680,17 +679,27 @@ object DeltaTable {
-   * Check if the provided `identifier` string, in this case a file path,
-   * is the root of a Delta table using the given SparkSession.
+   * Check if the provided `identifier` string represents a Delta table.
+   * Identifier strings can be either a path or a table name.
```
nit: table or view name

mmengarelli replied:
Roger

allisonport-db commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -680,17 +679,27 @@ object DeltaTable {
    * @since 0.4.0
    */
   def isDeltaTable(sparkSession: SparkSession, identifier: String): Boolean = {
+    // Try table name first
```
So I think what we should do is basically have two results, `isDeltaTableName` & `isDeltaTablePath` and then return `isDeltaTableName || isDeltaTablePath` basically (don't necessarily have to do exactly that just the flow of the reasoning.) `isDeltaTableName` would be false if the `identifier` is an invalid table name.

The way it's currently implemented gives the wrong result in the scenario that I have a delta table in a `identifier` dir in my current directory. Providing it's relative path here will be false, but prior to this change it would be true.

allisonport-db commented on core/src/test/scala/io/delta/tables/DeltaTableSuite.scala:
```diff
@@ -160,6 +160,14 @@ class DeltaTableSuite extends QueryTest
     }
   }
 
+  test("isDeltaTable - with table name") {
```
add test for when defined by both name and path such that exists=true. add test for when only defined by path, but in current directory and provide rel path (counter example provided above)

mmengarelli replied:
> add test for when defined by both name and path such that exists=true

This has been added

mmengarelli replied:
> add test for when only defined by path, but in current directory and provide rel path

I am seeing that save("path") - requires an absolute path and using option("path", "relpath") is placing the managed data under /hive/warehouse. Not relative to curr dir.

mmengarelli commented:
@allisonport-db Got it. I will make these changes

mmengarelli commented:
hi @tdas can I get a review of this? Thanks

allisonport-db reviewed (approved):
lgtm for these changes.

Only thing that just occurred to me, should isDeltaTable by _name_ be a separate API from isDeltaTable by _path_? I just realized I'm not sure we considered this, and not 100% sure this change is in line with the current API with `forPath` and `forName` + the ambiguity it adds. @tdas what do you think?

tdas commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -703,37 +704,52 @@ object DeltaTable {
    *
    * @since 0.4.0
    */
   def isDeltaTable(sparkSession: SparkSession, identifier: String): Boolean = {
```
does `dbName`.`tableName` format work? I would like to see tests for this.

tdas commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -703,37 +704,52 @@ object DeltaTable {
-   * Check if the provided `identifier` string, in this case a file path,
-   * is the root of a Delta table using the given SparkSession.
+   * Check if the provided `identifier` string represents a Delta table.
+   * Identifier strings can be either a path, table or view name.
```
i dont see any tests with view name.

tdas commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -703,37 +704,52 @@ object DeltaTable {
+    val isDeltaTableByName =
+      tableId.isSuccess && DeltaTableUtils.isDeltaTable(sparkSession, tableId.get)
+
+    val isDeltaTableByPath: Boolean = {
```
you dont need to compute this if `isDeltaTableByName` is true already .. right? 
so it might better to make this lazy .. either a lazy val or a def.

tdas commented on core/src/main/scala/io/delta/tables/DeltaTable.scala:
```diff
@@ -703,37 +704,52 @@ object DeltaTable {
-      DeltaTableUtils.isDeltaTable(sparkSession, identifierPath)
+    val tableId: Try[TableIdentifier] =
+      Try(sparkSession.sessionState.sqlParser.parseTableIdentifier(identifier))
+    val isDeltaTableByName =
```
isnt it possible to have a conflict where both a string can be interpreted as both a valid path (relative path with a scheme like s3:// ) and table name? In which case, you are deciding to resolve it by table first.. is that right?

tdas commented:
I had missed it but I think I agree with @allisonport-db as well.. i think its better to deprecate this API and create 2 new APIs for path and name separately. This will make it all less ambiguous.

mmengarelli commented:
> I had missed it but I think I agree with @allisonport-db as well.. i think its better to deprecate this API and create 2 new APIs for path and name separately. This will make it all less ambiguous.

So to confirm @tdas @allisonport-db : 
1. Revert changes to the following (and corresponding additional unit tests in DeltaTableSuite)
isDeltaTable(SparkSession, String)
isDeltaTable(sparkSession: SparkSession, identifier: String)

2. Deprecate both above methods in favor of:
isDeltaTableForName(sparkSession: SparkSession, tableOrViewName: String): Boolean
isDeltaTableForName(tableOrViewName: String): Boolean
isDeltaTableForPath(sparkSession: SparkSession, path: String): Boolean
isDeltaTableForPath(path: String): Boolean

3. Add corresponding unit test cases
isDeltaTableForName(tableOrViewName: String): Boolean
isDeltaTableForPath(path: String): Boolean

mmengarelli commented:
Proposing the following
Updated **DeltaTable.scala** 
Deprecated the following and delegated to the new corresponding isDeltaTableByPath() methods
- isDeltaTable(SparkSession, String)
- isDeltaTable(sparkSession: SparkSession, identifier: String)

Added:
- isDeltaTableByPath(sparkSession: SparkSession, path: String): Boolean
- isDeltaTableByPath(path: String): Boolean
- isDeltaTableByName(sparkSession: SparkSession, tableOrViewName: String): Boolean
- isDeltaTableByName(tableOrViewName: String): Boolean

Updated **DeltaTableSuite.scala**
Note: Left original `test("isDeltaTable ... ")` unit tests intact
Wanted to keep original unit tests intact until isDeltaTable() is finally deprecated.

Added the same unit tests to provide the same coverage to the new method (isDeltaTableByPath)
Added Corresponding unit tests (mirrored existing isDeltaTable tests):
- test("isDeltaTableByPath - with _delta_log dir") 
- test("isDeltaTableByPath - with empty _delta_log dir") 
- test("isDeltaTableByPath - with no _delta_log dir") 
- test("isDeltaTableByPath - with non-existent dir") 
- test("isDeltaTableByPath - with non-Delta table path") 
- test("isDeltaTableByPath - with table name in current path") 

Added new unit tests to cover isDeltaTableByName()
- test("isDeltaTableByName - with table name")
- test("isDeltaTableByName - with fully qualified table name")
- test("isDeltaTableByName - with view name")

mmengarelli commented:
Note: I annotated the new methods with  `@since 2.2.0` is that the intended version?

allisonport-db commented on core/src/test/scala/io/delta/tables/DeltaTableSuite.scala:
```diff
@@ -168,6 +168,74 @@ class DeltaTableSuite extends QueryTest
     }
   }
 
+  test("isDeltaTableByPath - with _delta_log dir") {
+    withTempDir { dir =>
+      testData.write.format("delta").save(dir.getAbsolutePath)
+      assert(DeltaTable.isDeltaTableByPath(dir.getAbsolutePath))
+    }
+  }
+
+  test("isDeltaTableByPath - with empty _delta_log dir") {
+    withTempDir { dir =>
+      new File(dir, "_delta_log").mkdirs()
+      assert(!DeltaTable.isDeltaTableByPath(dir.getAbsolutePath))
+    }
+  }
+
+  test("isDeltaTableByPath - with no _delta_log dir") {
+    withTempDir { dir =>
+      assert(!DeltaTable.isDeltaTableByPath(dir.getAbsolutePath))
+    }
+  }
+
+  test("isDeltaTableByPath - with non-existent dir") {
+    withTempDir { dir =>
+      JavaUtils.deleteRecursively(dir)
+      assert(!DeltaTable.isDeltaTableByPath(dir.getAbsolutePath))
+    }
+  }
+
+  test("isDeltaTableByPath - with non-Delta table path") {
+    withTempDir { dir =>
+      testData.write.format("parquet").mode("overwrite").save(dir.getAbsolutePath)
+      assert(!DeltaTable.isDeltaTable(dir.getAbsolutePath))
+    }
+  }
```
These are the same tests as we use on the deprecated API--can we combine them?

mmengarelli replied:
@allisonport-db  Shall I just deprecate these and have them delegate to the corresponding new ones?

allisonport-db replied:
we should keep the existing tests, deprecated doesn't mean we can break it

allisonport-db replied:
So what I meant is since these are duplicates what we can do is for each test do something like this:

```
  test("isDeltaTable - path - with _delta_log dir") {
    withTempDir { dir =>
      testData.write.format("delta").save(dir.getAbsolutePath)
      assert(DeltaTable.isDeltaTable(dir.getAbsolutePath))
      assert(DeltaTable.isDeltaTableByPath(dir.getAbsolutePath))
    }
  }
```
This way we don't have duplicate tests for each API

allisonport-db commented on core/src/test/scala/io/delta/tables/DeltaTableSuite.scala:
```diff
@@ -168,6 +168,74 @@ class DeltaTableSuite extends QueryTest
+  test("isDeltaTableByPath - with non-Delta table path") {
+    withTempDir { dir =>
+      testData.write.format("parquet").mode("overwrite").save(dir.getAbsolutePath)
+      assert(!DeltaTable.isDeltaTable(dir.getAbsolutePath))
```
also you missed adding `ByPath` here

allisonport-db commented on core/src/test/scala/io/delta/tables/DeltaTableSuite.scala:
```diff
@@ -168,6 +168,74 @@ class DeltaTableSuite extends QueryTest
+    }
+  }
+
+  test("isDeltaTableByPath - with table name in current path") {
+    val currDir = System.getProperty("user.dir")
+    try {
+      testData.write.format("delta").save(currDir + "/deltaTableName")
+      assert(DeltaTable.isDeltaTableByPath("deltaTableName"))
+    } finally {
+      JavaUtils.deleteRecursively(new File(currDir, "deltaTableName"))
+    }
+  }
```
With the changes this test is just testing a relative path doesn't have anything to do with table name

mmengarelli replied:
Remove?

allisonport-db replied:
update the test name

allisonport-db commented on core/src/test/scala/io/delta/tables/DeltaTableSuite.scala:
```diff
@@ -168,6 +168,74 @@ class DeltaTableSuite extends QueryTest
+    }
+  }
+
+  test("isDeltaTableByName - with table name") {
```
can we add a false case for isDeltaTableByName? maybe a (a) parquet table (b) non-existent table or view?

allisonport-db reviewed (commented):
code changes look good. left some comments on the tests

allisonport-db commented on core/src/test/scala/io/delta/tables/DeltaTableSuite.scala:
```diff
@@ -168,6 +168,75 @@ class DeltaTableSuite extends QueryTest
+    }
+  }
+
+  test("isDeltaTableByName - with view name") {
```
can we do a positive case for view?

allisonport-db reviewed (changes requested):
minor test changes. clarified a previous request. thanks!

allisonport-db commented on core/src/test/scala/io/delta/tables/DeltaTableSuite.scala:
```diff
@@ -168,6 +168,75 @@ class DeltaTableSuite extends QueryTest
+    }
+  }
+
+  test("isDeltaTableByName - with view name") {
```
can we add a positive case with view?

allisonport-db replied:
Actually this negative test doesn't really test anything; all `withView` does is drop `viewName` after the test. Just change this to a positive test (you can kind of copy the test above for a temp view with forName)
